contracts/LiquidityMiningCampaign.sol

```diff
@@ -0,0 +1,224 @@
+
+		if (user.amountStaked == 0) {
+			return;
+		}
+
+		updateUserAccruedReward(msg.sender); // Update the accrued reward for this specific user
+
+		LockScheme(_lockScheme).exit(msg.sender);
+		claimBonus(_lockScheme, msg.sender);
+		_claim(msg.sender);
+		stakingToken.safeApprove(transferTo, user.amountStaked);
+
+		StakeReceiver(transferTo).delegateStake(msg.sender, user.amountStaked);
+
+		totalStaked = totalStaked.sub(user.amountStaked);
+		user.amountStaked = 0;
+
+		for (uint256 i = 0; i < rewardsTokens.length; i++) {
+			user.tokensOwed[i] = 0;
+			user.rewardDebt[i] = 0;
+		}
+	}
+
+
+	function exitAndStake(address _stakePool, address _lockScheme) public onlyWhitelistedReceiver(_stakePool) nonReentrant{
+			
+		UserInfo storage user = userInfo[msg.sender];
+		
+		updateRewardMultipliers(); // Update the accumulated multipliers for everyone
+
+		if (user.amountStaked == 0) {
+			return;
+		}
+
+		updateUserAccruedReward(msg.sender); // Update the accrued reward for this specific user
+		
+		updateRewardMultipliers();
+        updateUserAccruedReward(msg.sender);
+			//todo check how to secure that 0 is the albt
+			uint256 finalRewards = user.tokensOwed[0].sub(userAccruedRewads[msg.sender]);
+			uint256 userBonus;
+			for (uint256 i = 0; i < lockSchemes.length; i++) {
+
+				uint256 additionalRewards = calculateProportionalRewards(msg.sender, finalRewards, lockSchemes[i]);
+				if(additionalRewards > 0) {
+
+				LockScheme(lockSchemes[i]).updateUserAccruedRewards(msg.sender, additionalRewards);
+				}
+				userBonus = userBonus.add(LockScheme(lockSchemes[i]).getUserBonus(msg.sender));
+			}
+
+		LockScheme(_lockScheme).exit(msg.sender);
+
+		uint256 amountToStake = getUserAccumulatedReward(msg.sender, 0).add(userBonus);
+		_exit(msg.sender);
+
+		stakingToken.safeApprove(_stakePool, amountToStake);
+
+		StakeReceiver(_stakePool).delegateStake(msg.sender, amountToStake);
+
+	}
+
+
+	function exit() public override {
+		revert("exit:cannot exit from this contract. Only exit and Unlock.");
+	}
+
+	function claim() public override {
+		revert("claim:cannot claim from this contract. Only exit and Unlock.");
+	}
+
+	function checkBonus(address _lockScheme, address _userAddress) public view returns(uint256 ) {
+		uint256 userBonus =  LockScheme(_lockScheme).getUserBonus(_userAddress);
+
+		return userBonus;
+	}
+
+	function claimBonus(address _lockScheme, address _userAddress) internal {
+
+		uint256 userBonus = checkBonus(_lockScheme, _userAddress);
+
+		IERC20Detailed(rewarsToken).safeTransfer(address(msg.sender), userBonus);
+
+		emit BonusTransferred(msg.sender, userBonus);
+	
+	}
+
+	function calculateProportionalRewards(address _userAddress, uint256 _accruedRewards, address _lockScheme) internal returns (uint256) {
+			
+			uint256 userStakedBalance = LockScheme(_lockScheme).getUserStakedBalance(_userAddress);
+
+			if(totalStaked > 0) {
+				uint256 ratio = userStakedBalance.div(totalStaked);
```

Comment:
Due to integer limitations, it is much better to do `userStakedBalance.mul(userStakedBalance).div(totalStaked)`